aggregator/aggregator.go

```diff
@@ -169,14 +171,25 @@ func (a *Aggregator) Shutdown() {
 	a.wg.Wait()
 }
 
-func (a *Aggregator) AddMaybe(buf [][]byte, val float64, ts uint32) {
-	if a.PreMatch(buf[0]) {
-		a.in <- msg{
-			buf,
-			val,
-			ts,
+func (a *Aggregator) AddMaybe(buf [][]byte, val float64, ts uint32) bool {
+	if !a.PreMatch(buf[0]) {
+		return false
+	}
+
+	if a.DropRaw {
+		_, ok := a.matchWithCache(buf[0])
+		if !ok {
+			return false
 		}
 	}
+
+	a.in <- msg{
+		buf,
+		val,
+		ts,
+	}
+
+	return a.DropRaw
```

Comment:
the way it is in master is because:
* table.Dispatch may not block, it needs to be fast, so anything it calls synchronously (e.g. AddMaybe) needs to be fast
* so AddMaybe only does cheap checks (in PreMatch) and if those can't reliably determine the metric doesn't match, AddMaybe feeds it into the `.in` channel which can be processed async (and do slower operations like the regex matching)

If we're going to call matchWithCache here (and do a full match check) we block the table, which is expensive especially when the cache is cold or when disabled.
Of course the larger problem here is we want AddMaybe to respond immediately with whether the msg can be dropped in the table, but at the same time, we want its processing not to block the table...
I think that's the crux of this PR.

Perhaps we should rearchitect this, and make it blocking, in which case we can do away with the 2 separate steps of the prematch and the real match, just do one -if-match-then-add operation which blocks, but that means we should run table.Dispatch concurrently and put a buffer in front of it.

Reply:
that approach would probably make the entire pipeline more performant and elegant. seems like doing the buffering and concurrency in 1 place (table.Dispatch), enables us to replace lots of buffered channels with direct function calls and replace 2 step matching with single steps in more than just the aggregators. also the routes and destinations would benefit I think.
but this needs a bit more study

Reply:
Yeah, that's why I tried to structure this to have as small an impact as possible, the only time the behavior changes is when the aggregator has dropRaw enabled, and it only blocks for the final regex match when the prematch has already passed, so hopefully that should be a scenario where the cache will be able to operate efficiently and minimize the amount of time spent blocking the table.

Reply:
I missed that the matchWithCache is only called in case of dropRaw, that indeed makes my concerns mostly irrelevant for now.  I think this PR is good, i will just add a note that enabling dropRaw may slow down the relay.